### ISSUE TITLE
[HOLD for payment 2023-12-06] [$1000] Web - Pressing ESC on attachment preview glitches the preview

### ISSUE THREAD

**Issue by kbecciv** [Expensify/App#27237]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Open the [app](https://staging.new.expensify.com/r/1000398671829417)
2. Click on any chat
3. Click '+' add attachment
4. Attach multiple images
5. Open first uploaded image
6. Click on ESC `multiple times`, observe that preview closes, reopens with different image and closes

## Expected Result:
 App should close preview on ESC click (note you need to press Esc multiple times)

## Actual Result:
 App glitches on closing preview on ESC click

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android / native
- [ ] Android / Chrome
- [ ] iOS / native
- [ ] iOS / Safari
- [x] MacOS / Chrome / Safari
- [ ] MacOS / Desktop

**Version Number:** 1.3.61-1
**Reproducible in staging?:** y
**Reproducible in production?:** y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation

https://github.com/Expensify/App/assets/93399543/1c6168d6-ff13-4892-892f-4a98015c026a

https://github.com/Expensify/App/assets/93399543/9e5b35a2-db33-474a-957e-6fe952084b2e

**Expensify/Expensify Issue URL:**
**Issue reported by:** @dhanashree-sawant
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1693558112570809

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01ebd6c095f0ad277c</li>
        <li>Upwork Job ID: 1701562714140925952</li>
        <li>Last Price Increase: 2023-10-03</li>
    </ul>
</details>

**Comment by ishpaul777** [comment]
## Proposal
### Problem Statement
Pressing ESC Attachement preview glitches the preview.

### Root Cause
The root cause is when keyboard shortcut is triggered we call Modal.close(() => callback), here modal.close is dependent on the onClose prop which we pass in Modal component. But in this case we have 2 Modals in stack of AttachmentsModal and other is ConfirmModal used for invalid file type. so variable we are using to store onClose prop is getting overridden by ConfirmModal onClose prop (which is null if modal not visble)
### Solution
Conditionally rendering the ConfirmModal based on the whether there is a validation error or not, and Modifying onModalHide prop passed to AttachementModal will solve this issue.
```js
// AttachmentsModal.js
{
  isAttachmentInvalid && (
      <ConfirmModal
          title={attachmentInvalidReasonTitle ? translate(attachmentInvalidReasonTitle) : ''}
          onConfirm={closeConfirmModal}
          onCancel={closeConfirmModal}
          isVisible={isAttachmentInvalid}
          prompt={attachmentInvalidReason ? translate(attachmentInvalidReason) : ''}
          confirmText={translate('common.close')}
          shouldShowCancelButton={false}
      />
  )
}
```
```js
// ReportAttachments.js
          <AttachmentModal
            onModalHide={() => {
                // if active route is not the attachment route, user maybe navigated to search using KeyboardShortcut Cmd+K/Cmd+Shift+K below can be moved to dismissModal function also
                const activeRoute = Navigation.getActiveRoute();
                if (activeRoute.includes('/attachment')) {
                    Navigation.dismissModal(reportID)
                }
            }}
            /// other props
          />
````
### Alternative Solution
As this problem can be solved by conditionally rendering the ConfirmModal. There is a downside to the approach we are following as explained in root cause, we can only have one modal in Stack at a time. So where there is multiple modals in stack, we have to conditionally render them.
A better approach would be to use a Array to store the onClose, and setOnClose function will push the onClose function to the array along with a ModalID. When unmounting the modal will remove the onClose function from the array. This way we can have multiple modals in stack and onClose will be called in the order of modals in stack.
Pseudo Code
-  Modal Mounted setOnClose({ onClose, modalID })
-  setOnClose will push the onClose function to the array along with a ModalID
-  Modal Unmounted removeOnClose(modalID)

**Comment by zukilover** [comment]
## Proposal
### Please re-state the problem that we are trying to solve in this issue.
App glitches on closing preview on ESC click

### What is the root cause of that problem?
`AttachmentModal` is using unnecessary `useCallback` for `closeModal` function. This causes React to not throw away the cached function.

https://github.com/Expensify/App/blob/5f668c669a501a94caebbb434a656580e3bcfb24/src/components/AttachmentModal.js#L307-L309

### What changes do you think we should make in order to solve the problem?
Remove the `useCallback` wrapper for `closeModal`, e.g.:
```
    const closeModal = () => {
        setIsModalOpen(false);
    };
```

This is the safest and optimal way since the modal visibility is controlled by a local state.

### What alternative solutions did you explore? (Optional)
N/A

**Comment by ishpaul777** [comment]
While my solution is solving the issue I think the root cause is correctly identified by @zukilover, the solution proposed by me is solving the issue of Navigation issue when CMD+K or CMD+SHIFT+K is used for navigation (reported by a contributor on slack already) I am marking my off-topic (unless we decide to solve this issue alog with these).

**Comment by Christinadobrzyn** [comment]
Not sure why @adelekennedy and I are both assigned but I did the testing and can reproduce this only in Chrome. Looks like External is already added for some proposals.

**Comment by AmjedNazzal** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

Web - Pressing ESC on attachment preview glitches the preview 

### What is the root cause of that problem?

The root of this issue is that we are nesting ComposerWithSuggestions inside of AttachmentModal and then we are using a useEffect to focus the composer without a delay which causes triggering focusWithDelay while AttachmentModal is still mounted resulting in it re-rendering and causing the observed glitch

https://github.com/Expensify/App/blob/9675a77fc073b6197dd844b29e4ac27faf1e480f/src/pages/home/report/ReportActionCompose/ComposerWithSuggestions.js#L433-L435

### What changes do you think we should make in order to solve the problem?

The simplist solution we can introduce is by actually delaying the focus on the composer so that AttachmentModal has enough time to unmount and wouldn't re-render again.

```javascript
focus(true);
}, [focus, prevIsFocused, prevIsModalVisible, isFocused, modal.isVisible, isNextModalWillOpenRef]);
```

### Result

https://github.com/Expensify/App/assets/74202751/38b7c9b9-9261-4da0-93db-fe804fdb4a31

**Comment by astrohunter62** [comment]
## Proposal
### Please re-state the problem that we are trying to solve in this issue.
Pressing or double pressing ESC on the attachment preview glitches the preview
### What is the root cause of that problem?
- The AttachmentModal unnecessarily utilizes useCallback for the closeModal function. This prevents React from discarding the cached function.
- There is a focus issue when displaying the attachment preview in the modal. The current image in the attachment preview modal is not visually indicated as focused.
### What changes do you think we should make in order to solve the problem?
We can remove the useCallback wrapper for the closeModal function. And add focus when displaying the image preview.
For example:
src/components/imageView/index.js
```
    useEffect(() => {
        if (imageRef) {
            imageRef.current.focus();
            return imageRef.current.blur();
        }
    }, []);
    ...
   <PressableWithoutFeedback
    ...
      ref={imageRef}
   >
```
- Result

https://github.com/Expensify/App/assets/136574831/5821d38d-5afe-4e1f-8355-f6627ce98e15

https://github.com/Expensify/App/assets/136574831/ac4a25cf-3a3c-478f-a583-6139942346d3

**Comment by astrohunter62** [comment]
Contributor details
Your Expensify account email: astro.dev131@gmail.com
Upwork Profile Link: https://www.upwork.com/freelancers/~01bfd69d6e1c5de05c

**Comment by Christinadobrzyn** [comment]
I think we're reviewing proposals!

**Comment by Christinadobrzyn** [comment]
hi @abdulrahuman5196! Can you let me know what you think of the current proposals and if you'd like to see more?

**Comment by Christinadobrzyn** [comment]
Hey @abdulrahuman5196 would you like to see more proposals?

**Comment by Christinadobrzyn** [comment]
reached out to @abdulrahuman5196 to see if we can get a review of these

**Comment by mvtglobally** [comment]
Issue not reproducible during KI retests. (First week)

**Comment by abdulrahuman5196** [comment]
Reviewing now

**Comment by abdulrahuman5196** [comment]
@Christinadobrzyn I am not able to repro the issue in the latest main. Could you kindly check again if reproducible?

**Comment by Christinadobrzyn** [comment]
Yep! This is no longer happening. I'll ask QA to test just to make sure.

**Comment by kavimuru** [comment]
@Christinadobrzyn I am not able to reproduce

https://github.com/Expensify/App/assets/43996225/f897a7d4-410c-433d-913c-f332fa8c137e

**Comment by astrohunter62** [comment]
Hi, @Christinadobrzyn, did you press ECS twice on the first attachment preview?

**Comment by Christinadobrzyn** [comment]
QA confirmed this isn't happening anymore. Going to close!

**Comment by Christinadobrzyn** [comment]
Ah, it is still happening when you press Esc multiple times... I'll update the OP.

**Comment by astrohunter62** [comment]
I updated my proposal https://github.com/Expensify/App/issues/27237#issuecomment-1719794579

**Comment by abdulrahuman5196** [comment]
Rechecking this again.

**Comment by abdulrahuman5196** [comment]
Got it. able to repro the issue if Esc is pressed multiple times

**Comment by abdulrahuman5196** [comment]
On @zukilover 's proposal here https://github.com/Expensify/App/issues/27237#issuecomment-1715649797 I am not sure of the root cause, but i tried the solution. It is not solving the issue.

On @astrohunter62 's proposal here https://github.com/Expensify/App/issues/27237#issuecomment-1719794579. I not sure if the root cause is correct. But the solution suggestion like a workaround to not show FlatList on non focused state to avoid the issue. So I am not inclined to the solution.

@Christinadobrzyn Currently there is no proposal to approve, can we increase the bounty given that the issue is couple of weeks old already.

**Comment by astrohunter62** [comment]
@abdulrahuman5196 
I believe the root cause is valid. When I press the 'ESC' key, it triggers the 'onClose' function in the modal. If I click 'ESC' before the 'onModalHide' event is triggered, the 'AttachmentCarousel' is re-rendered with the last item as the entry item. Consequently, the modal experiences a glitch.

**Comment by Christinadobrzyn** [comment]
Increased the bounty on this one to get more proposals.

**Comment by ilkeruyanik** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

When the Attachment Carousel is open, pressing the "ESC" key multiple times causes the modal to close and then reopen.

### What is the root cause of that problem?
The `SidebarLinks` component has an event listener for the "ESC" key, but it doesn't check whether the modal is currently visible or not. When the "ESC" key is pressed, it triggers the `Navigation.dismissModal()` function with check `modal.willAlertModalBecomeVisible`.

https://github.com/Expensify/App/blob/4662f62e8905a969909977a87ebc64cfb72f6d09/src/pages/home/sidebar/SidebarLinks.js#L96-L104

Additionally, the `AttachmentModal` also calls `Navigation.dismissModal()` when it hides modal.

https://github.com/Expensify/App/blob/4662f62e8905a969909977a87ebc64cfb72f6d09/src/pages/home/report/ReportAttachments.js#L33

1. On first "ESC" press, modal is closed by `BaseModal` component close modal with `onBackButtonPress` prop, and `AttachmentModal` call `Navigation.dismissModal()`.
2. On second "ESC" press, the `SidebarLinks` component's event listener trigger `Navigation.dismissModal()`.

This leads to two consecutive calls to `Navigation.dismissModal()`, causing the issue. 

Furthermore, the `BaseModal` component only closes the modal with the "ESC" key using its `onBackButtonPress` prop, but it doesn't update the modal's visibility status using `Modal.setModalVisibility(false)`. This action write Onyx to modal's visibility.

### What changes do you think we should make in order to solve the problem?
To prevent multiple calls to `Navigation.dismissModal()`, we will check modal visibility in `SidebarLinks` component with
```
if(modal.isVisible){
    Navigation.dismissModal();
}
```
Also, we will add `Modal.setModalVisibility(false)` to `BaseModal` `onBackButtonPress` prop.

```
onBackButtonPress={() => {
    Modal.setModalVisibility(false);
    onClose();
}}
```

By implementing these two changes, we will prevent multiple calls to `Navigation.dismissModal()`. The `onBackButtonPress` action will set the modal's visibility to false when the modal is closed  with the "ESC" key, ensuring that it won't be called again in the `SidebarLinks` component.

### What alternative solutions did you explore? (Optional)

N/A

### Result

[result.webm](https://github.com/Expensify/App/assets/93402058/abe6acd3-a5b8-4911-96f2-d055fbc5c892)

**Comment by shubham1206agra** [comment]
@bernhardoj Can you confirm https://github.com/Expensify/App/issues/23959 will fix this too?

**Comment by shubham1206agra** [comment]
@bernhardoj For me, it is pretty related cause when we move the modalHide logic to onDismiss, we will not have race condition for `Navigation.dismissModal()` in SidebarLinks and ReportAttachment
See @ilkeruyanik proposal

**Comment by shubham1206agra** [comment]
I have already asked the team which is bumping the version of RNWeb. Lets test again after the version bump

**Comment by abdulrahuman5196** [comment]
Thank you @bernhardoj for 

>I read @ilkeruyanik proposal and I can confirm double Navigation.dismissModal is causing the issue, but I still don't know why double Navigation.dismissModal reopens the attachment carousel. I think it's better to focus on that issue.

I agree on the same.

**Comment by abdulrahuman5196** [comment]
@ilkeruyanik and @tsa321 both  proposals mentions the issue happens on `Navigation.dismissModal()` being called twice and providing solutions to make sure only one call is made. But the root cause would need more information on why `Navigation.dismissModal()` causes the attachment viewer to show the next image and close. Without knowing this, if we only put a solution to make sure only one call is made, we would be putting a workaround solution. 

So I would kindly request the contributors to recheck the root cause and add more information on the same.

**Comment by ilkeruyanik** [comment]
@abdulrahuman5196 

I investigated further this issue, and I found;

When user press ESC two times, the `SidebarLinks` component's event listener trigger `Navigation.dismissModal()` as I mentioned in proposal. 
However, at that moment, the modal doesn't actually disappear.  It cause trigger `onViewableItemsChanged` in `AttachmentCarousel`. 

Because we've set the `Navigate.route` in the `onCarouselAttachmentChange` prop in the `ReportAttachments` component, the `AttachmentCarousel` component's `onViewableItemsChanged` event also triggers the `Navigate.route`

https://github.com/Expensify/App/blob/389d7b0c4b96c6f2a2295055c685791d35b65252/src/pages/home/report/ReportAttachments.js#L34-L37

 Note that  in this case `viewableItems` parameters are changed, up to rendered items. You can test it with set props `initialNumToRender={1} windowSize={1}  maxToRenderPerBatch={1}` on FlatList in `AttachmentCarousel`. It explains why this bug is not happen when carousel has only one image. Because it has only one rendered element.

On the other hand,  `onModalHide` is triggered in `AttachmentModal` component by `onBackButtonPress` props in the `BaseModal` component. It call `Navigation.dismissModal()` second time and return report page.

In summary, the root cause is `Sidebarlinks` components event listener trigger `Navigation.dismissModal()` while modal is not disappear. It cause triggers `onViewableItemsChanged`. It call `onCarouselAttachmentChange`, finally calls `navigate.route`.

When we do implication of my proposal, event listener in `Sidebarlinks` will not call `Navigation.dismissModal()`. It will solve this issue.

**Comment by pradeepmdk** [comment]
### Proposal
### Please re-state the problem that we are trying to solve in this issue.
when pressing esc key double time attachement preview glitches 

### What is the root cause of that problem?
https://github.com/Expensify/App/blob/389d7b0c4b96c6f2a2295055c685791d35b65252/src/pages/home/sidebar/SidebarLinks.js#L81
when pressing esc key `SidebarLinks` will called first and react native modal onBackButtonPress in baseModal  is called second.
so at the time   `willAlertModalBecomeVisible` set false so second time pressing esc key `SidebarLinks` Navigation.dismissModal(); will triggered.
but in the mean time in `AttachmentModal -> AttachmentCarousel`  we are using flatlist on that we are using `onViewableItemsChanged` it will triggered based on the `viewabilityConfig` 
 
https://github.com/Expensify/App/blob/389d7b0c4b96c6f2a2295055c685791d35b65252/src/pages/home/report/ReportAttachments.js#L34-L37

whenever `onViewableItemsChanged` we are navigating `REPORT_ATTACHMENTS`.
so when pressing double esc before completing the modal close and open it again. 
in basemodal  when `onModalHide={hideModal}` called only modal will removed completely.
https://www.npmjs.com/package/react-native-modal 
so that glitches happened.

### What changes do you think we should make in order to solve the problem?

we to prevent the `onViewableItemsChanged` 
options 1) 
https://github.com/Expensify/App/blob/389d7b0c4b96c6f2a2295055c685791d35b65252/src/components/Attachments/AttachmentCarousel/index.js#L24
here we need to add 
`waitForInteraction: true,`
https://reactnative.dev/docs/flatlist#viewabilityconfig
 
options2) 

https://github.com/Expensify/App/blob/389d7b0c4b96c6f2a2295055c685791d35b65252/src/components/AttachmentModal.js#L403
here
`!_.isEmpty(props.report) && isModalOpen`

**Comment by paultsimura** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
When clicking "Esc" multiple times, the attachment preview glitches.

### What is the root cause of that problem?
_The root cause here is buried quite deep and required a lot of debugging, so I'll try to be as detailed as possible here._

#### TL;DR
When we are hiding the attachments modal, we try to focus the composer, which causes the `onScroll` event on the `FlatList` inside the `AttachmentCarousel`. This causes an avalanche effect of bugs.

This same RCA can be easier reproduced here: https://github.com/Expensify/App/issues/29109

#### More detailed explanation:

In ComposerWithSuggestions, we are trying to focus the `Composer` right after we hide the modal:

https://github.com/Expensify/App/blob/01ef03f850047ca36caa21fd38b9f2f8555ef78d/src/pages/home/report/ReportActionCompose/ComposerWithSuggestions.js#L481

However, since the modal dismissal is an animated process, it's not happening right away. So we are, in fact, focusing on the composer when the modal is still open. The focus action fires a scroll event, which is then captured by the FlatList inside the AttachmentsCarousel (since FlatList is a combination of a VirtualizedList and a ScrollView).

Here's the call stack that comes all the way back to the ComposerWithSuggestions:
<img width="376" alt="image" src="https://github.com/Expensify/App/assets/12595293/975eb431-7abe-4526-9512-7a79fc6651cb">

<img width="376" alt="image" src="https://github.com/Expensify/App/assets/12595293/a8ed16c1-999b-4aca-a9c4-f0b8d82382f5">

This, as a result, causes the `onViewableItemsChanged` of the FlatList, which causes glitchy change of the image in this issue, and the endless redirection loop in https://github.com/Expensify/App/issues/29109.

### What changes do you think we should make in order to solve the problem?

**First,** we need to call the composer focus on modal hide **with delay**:

```diff
-       focus();
+       focus(true);
    }, [focus, prevIsFocused, prevIsModalVisible, isFocused, modal.isVisible, isNextModalWillOpenRef]);
```

https://github.com/Expensify/App/blob/01ef03f850047ca36caa21fd38b9f2f8555ef78d/src/pages/home/report/ReportActionCompose/ComposerWithSuggestions.js#L469-L482

**Second,** the `focusWithDelay` relies on `ComposerFocusManager.isReadyToFocus()`:

https://github.com/Expensify/App/blob/3789cc36ed4c96ea4e94bf1552d87cc19b66beeb/src/libs/focusWithDelay.ts#L25-L30

This is, in fact, barely useful since the `ComposerFocusManager.resetReadyToFocus()` is called only in one place, and it's on Android:

https://github.com/Expensify/App/blob/11717c5772c631af6629d443270b8fbf107bf875/src/components/Modal/index.android.js#L12-L14

We should set the composer as "not ready to focus" every time a modal is open, and set it back to "ready to focus" when the modal is dismissed.

For this, we need to add the following prop to the `ReactNativeModal` component (it was by mistake removed in https://github.com/Expensify/App/pull/25672):

```diff
            onModalWillShow={() => {
                ComposerFocusManager.resetReadyToFocus();
            }}
```

This solution alone would be enough to solve the current issue. However, if we want to fix both this and https://github.com/Expensify/App/issues/29109 in one take, we should make the following change:

**Third,** we need to correctly handle setting back the composer to "Ready to focus".

Currently, the `ComposerFocusManager.setReadyToFocus()` is called in 2 places in `BaseModal`:

https://github.com/Expensify/App/blob/a348859ff43f38c487de96a3f1b4a75428f493b9/src/components/Modal/BaseModal.js#L123-L125

https://github.com/Expensify/App/blob/a348859ff43f38c487de96a3f1b4a75428f493b9/src/components/Modal/BaseModal.js#L75-L77

Calling it in the handleDismissModal was preferred because it's called finally, after the onModalHide. But this specific method is not called in case of a non-fullscreen modal, and it's not called on navigating back via browser navigation.

The non-fullscreen case is handled in `hideModal`, but since the Attachment modal is full-screen, the "go-back" navigation will not call the `ComposerFocusManager.setReadyToFocus()`, so the composer will not be focused on closing the attachment preview by navigating back.

That's why my suggestion is to move the `ComposerFocusManager.setReadyToFocus()` fully into `hideModal`:
Remove handleDismissModal and make the change:

```diff
    const hideModal = useCallback(
        (callHideCallback = true) => {
            if (shouldSetModalVisibility) {
                Modal.setModalVisibility(false);
            }
            if (callHideCallback) {
                onModalHide();
            }
            Modal.onModalDidClose();
-           if (!fullscreen) {
            ComposerFocusManager.setReadyToFocus();
-           }
        },
        [shouldSetModalVisibility, onModalHide, fullscreen],
    );
```

As a result, we will have both issues fixed.

### What alternative solutions did you explore? (Optional)
Optionally, we could only call the `ComposerFocusManager.setReadyToFocus()` for the attachment modal here:
https://github.com/Expensify/App/blob/7fb66cdd840ba14f5c958dd60fc1bb4f75d58bab/src/pages/home/report/ReportAttachments.js#L33

But I don't see any major reason against my step 3 of this proposal, as it will be more versatile.

**Comment by abdulrahuman5196** [comment]
The latest proposals have much information which needs dive deep. Will review again tomorrow.

**Comment by abdulrahuman5196** [comment]
Still investigating. Will close over the weekend.

**Comment by abdulrahuman5196** [comment]
Thank you everyone for the root cause analysis. 
Status so far: I was able to review the root cause which is `onViewableItemsChanged` of flatlist getting called causing this issue.
I am have checked couple of proposals. Need to do some different platform testing on some of the proposals, before providing more info. Will post here once that is done.

**Comment by abdulrahuman5196** [comment]
Still investigating. Will provide update sooner

**Comment by abdulrahuman5196** [comment]
RE-Reviewing now

**Comment by abdulrahuman5196** [comment]
From investigation, the proposal by @pradeepmdk here https://github.com/Expensify/App/issues/27237#issuecomment-1751743967 seems simpler fixes which is aiming to fix in the attachment component itself which is good compared to other proposals so far.
I tried 1st option and it didn't work for me. Just reducing the flicker time IMO.

2nd option is one I am more inclined for now. But it is showing loading indicator when closing the attachment modal.


https://github.com/Expensify/App/assets/46707890/f9590752-1735-4e54-8206-e43afd407e21



@pradeepmdk Could you check on these concerns and update the proposal if possible to fix the concerns?

**Comment by pradeepmdk** [comment]
@abdulrahuman5196 i updated propsoal. and i agree with @paultsimura my solution not solved https://github.com/Expensify/App/issues/29109 because when we press back button https://github.com/Expensify/App/blob/389d7b0c4b96c6f2a2295055c685791d35b65252/src/components/AttachmentModal.js#L370 it's not triggering. so now the decision with to you but   @paultsimura proposes solving both issues.
 
but `waitForInteraction: true,` worked on initially but now not working I am not sure what happened now 😞 

and one more thing but all are saying https://github.com/Expensify/App/blob/389d7b0c4b96c6f2a2295055c685791d35b65252/src/pages/home/report/ReportActionCompose/ComposerWithSuggestions.js#L481 focus on immediately cause this issue. then this issue is always there, right? if we have more images then when navigating between images then no issues on here only thing before navigating images we are facing issues. any idea on this?
ref: (from staging no issues)

https://github.com/Expensify/App/assets/16595846/38574fa3-2d3b-4243-bce7-8c85da009b3e

**Comment by abdulrahuman5196** [comment]
Hi folks thanks for the detailed discussion. 
Just to note simpler fixes are not always workaround fixes. But anyways since the proposal couldn't fix the loading indicator issue and other similar issue we have to drop it.

The proposals left currently to re-review are

1) @ilkeruyanik proposal here https://github.com/Expensify/App/issues/27237#issuecomment-1746045280 :
I am not aligned on this since the proposal is still attempting to present multiple dismissModal by using stored values. But anyways will compare with other proposals as well.

2) @tsa321 's proposal here https://github.com/Expensify/App/issues/27237#issuecomment-1746060535
Could you kindly update the proposal if it doesn't included all the changes you have mentioned.

3) @paultsimura 's proposal here https://github.com/Expensify/App/issues/27237#issuecomment-1753912527
Need to fresh check this one.

**Comment by abdulrahuman5196** [comment]
Re-reviewing now

**Comment by abdulrahuman5196** [comment]
Still on this phase going back and forth testing - https://github.com/Expensify/App/issues/27237#issuecomment-1779814093.
Will update in a day.

**Comment by Christinadobrzyn** [comment]
Hi @abdulrahuman5196 just checking to see if there's any update based on the testing. Thanks!

**Comment by abdulrahuman5196** [comment]
Hi, Will review today just back from weekend

**Comment by Christinadobrzyn** [comment]
hi @abdulrahuman5196 can you provide an update for us? Thanks!

**Comment by abdulrahuman5196** [comment]
Sorry for dragging on. Still investigating proposal. Will cover over weekend or reach out for another C+ on monday.

**Comment by shubham1206agra** [comment]
@abdulrahuman5196 If you want help from another C+, I will be happy to help you cause I have been following this issue from the start.

**Comment by abdulrahuman5196** [comment]
Sure. Let me check over this weekend, if not will forward.

**Comment by abdulrahuman5196** [comment]
All the 3 pending proposals are working well. Let me compare them and come back with results.

**Comment by abdulrahuman5196** [comment]
@paultsimura 's proposal here https://github.com/Expensify/App/issues/27237#issuecomment-1753912527 looks good and work well. It is better than other proposal IMO. Reason being, the main root cause is that `onViewableItemsChanged` is being called incorrectly from FlatList during our usage. But its not worth some upstream fix and since fix at our end is relatively small in all proposals. And another reason I think of is that if we add `Navigation.dismissModal` in future in another place, we don't have to do any extra checks specific to this issue.
And I think we can just implement the **First change** proposed, I think some of the other changes proposed have been already implemented with some other PRs.

🎀 👀 🎀
C+ Reviewed

**Comment by paultsimura** [comment]
Thank you, the PR will be ready today.

**Comment by Christinadobrzyn** [comment]
I'm going ooo until Monday Nov 27th so going to assign another BZ teammate to monitor this until I'm back. Thanks!

Status - reviewing PR! 🎉

**Comment by abdulrahuman5196** [comment]
@tsa321  let me check if you had proposed the solution first.

Cc: @paultsimura do share if you have thoughts

**Comment by abdulrahuman5196** [comment]
Looking at the proposals

@tsa321 edit the original proposal to include focus code suggestions only 3 weeks back https://github.com/Expensify/App/issues/27237#issuecomment-1746060535

Whereas @paultsimura had the suggestion on his original proposal in Oct 10 itself https://github.com/Expensify/App/issues/27237#issuecomment-1753912527 that let to my decision. 


But seems @tsa321 suggested the solution as part of analysis replies here https://github.com/Expensify/App/issues/27237#issuecomment-1751632831 which happened at Oct 8, 2 days before @paultsimura 's solution. 

So if we consider only main proposals @paultsimura would be one to choose. 
But I am not sure if we would consider replies and conversations which has code change suggestions as a valid proposals

@tsa321 / @paultsimura  let me know if I got the timelines in any way wrongly. 

@hayata-suenaga do you have any suggestions here?

And note for contributors: We have guideline to update original proposal everytime you think of a solution is worth mentioning to avoid this type of concerns.

**Comment by paultsimura** [comment]
@abdulrahuman5196 I agree that the proposal of @tsa321 was not updated on time. Because of this, I did not even see those suggested changes before posting my proposal.

But in order not to make a decision based only on this, here are a couple more thoughts:

Setting `focus(true)` is, in fact, only a partial solution, because without my third proposed change (which I still think we should implement), we have an issue: when clicking the browser "back" button, the composer does not get focused similar to clicking "x" or "esc".

> Third, we need to correctly handle setting back the composer to "Ready to focus".


https://github.com/Expensify/App/assets/12595293/594e7d75-1e0d-4cc8-817b-be699d33b027

@tsa321 did not include it in their [original discussion](https://github.com/Expensify/App/issues/27237#issuecomment-1751632831) but added it in the modified proposal **after** mine was posted, meaning they considered it valuable enough to be included in the proposal, but missed this change in the discussion.

Here is the behavior with the fix applied:

<details><summary>Fixed focus</summary>

https://github.com/Expensify/App/assets/12595293/0c82d48c-14d2-40ab-9d5c-9cbaf111954f

</details>

**Comment by paultsimura** [comment]
I'd just like to add that in the past, I had a similar situation when my proposal was rejected because as was originally added in a discussion. The reason was that "You should be confident about the proposed solution, so when you have it, it should be reflected in the proposal".

To me, it seems logical that when some new idea comes to mind, it should be reflected in the proposal, otherwise waiting for some extra response before updating the proposal means that the Contributor is not sure about the proposed solution and doesn't want to risk putting it in the proposal before an extra confirmation.

I would prefer to avoid going back and forth further here because it can lead to some personal confrontations. I'll let @abdulrahuman5196 and @hayata-suenaga make the further decision taking into account the timing and these 2 comments: 
https://github.com/Expensify/App/issues/27237#issuecomment-1817412283
https://github.com/Expensify/App/issues/27237#issuecomment-1817417013

**Comment by Christinadobrzyn** [comment]
I'm going ooo until Monday Nov 27th so going to assign another BZ teammate to monitor this until I'm back. Thanks!

Status: working on PR

**Comment by paultsimura** [comment]
Thanks. To keep things relatively fair for all, I'd like to offer $250 of my bounty to @tsa321 when the issue gets fixed.

**Comment by Christinadobrzyn** [comment]
This PR is in review so moving with GH to weekly.

**Comment by paultsimura** [comment]
@Christinadobrzyn the PR is already approved & merged 🤔

**Comment by Christinadobrzyn** [comment]
Sorry that was my lazy way of writing, "we're waiting for the PR to go into production so this can be paid".

**Comment by Christinadobrzyn** [comment]
prepping for payment 

Payouts due:

Issue Reporter: $250 @dhanashree-sawant (in Upwork)
Contributor: $1000 @paultsimura (in Upwork)
Contributor+: $1000 @abdulrahuman5196 (in Upwork)

Eligible for 50% #urgency bonus? N

Upwork job is [here](https://www.upwork.com/jobs/~01b2340a470e9fd6fe).

@abdulrahuman5196 do we need a regression for this?

**Comment by paultsimura** [comment]
Thank you @Christinadobrzyn 
Based on this comment: https://github.com/Expensify/App/issues/27237#issuecomment-1820835956, could you please split my bounty into $750 for me, and $250 for @tsa321?

**Comment by abdulrahuman5196** [comment]
>The PR that introduced the bug has been identified. Link to the PR:
The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake. Link to comment:
A discussion in #expensify-bugs has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner. Link to discussion:

Not a regression.

>Determine if we should create a regression test for this bug.

Yes.

>If we decide to create a regression test for the bug, please propose the regression test steps to ensure the same bug will not reach production again.

1) Attach multiple images in a chat
2) Open the first uploaded image
3) Click on ESC multiple times
4) Verify that the attachment modal doesn't reopen
5) Open an attachment again
6) Click the "Back" button in the browser
7) Verify that the composer gets focused
8) Open an attachment again
9) Click the "X" button on the modal
10) Verify that the composer gets focused

@Christinadobrzyn

**Comment by Christinadobrzyn** [comment]
Awesome! Thanks @abdulrahuman5196! 

I paid you and @paultsimura in Upwork based on this payment structure https://github.com/Expensify/App/issues/27237#issuecomment-1841947725

@tsa321 and @dhanashree-sawant can you accept the Upwork offer and I'll pay you?

**Comment by dhanashree-sawant** [comment]
Hi @Christinadobrzyn, sorry for the delay, the job was actually raised after 30 August but you have sent me 250$ offer, can you rectify it once to 50$?

**Comment by Christinadobrzyn** [comment]
Thanks @tsa321! I paid you in Upwork. 

Thanks for catching that @dhanashree-sawant  - I got my dates mixed up - I sent you a modified offer, would you mind accepting so I can pay you?

**Comment by dhanashree-sawant** [comment]
That's fine, I have accepted the offer.

**Comment by Christinadobrzyn** [comment]
Thanks @dhanashree-sawant! I paid you in Upwork based on our [payment summary.](https://github.com/Expensify/App/issues/27237#issuecomment-1841947725) 

This is good to close - thanks everyone!
